### PR TITLE
Sorry for this awful pull request

### PR DESCRIPTION
**Pull request by camelot729**
fix mistake and code style, add edit/destroy comments, add avatar to user

### CODE REVIEW

File: book_review/app/assets/javascripts/book_ratings.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (ghost) [inline book_review/app/assets/javascripts/book_ratings.coffee]: We prefer to not commit empty files to repository, pls remove it

File: book_review/app/assets/stylesheets/book_ratings.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the book_ratings controller here.
```
Comment (ghost) [inline book_review/app/assets/stylesheets/book_ratings.scss]: empty file too

File: book_review/app/controllers/book_ratings_controller.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
+class BookRatingsController < ApplicationController
+  before_action :set_book_rating, only: [:show, :edit, :update, :destroy]
+
+  # GET /book_ratings
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: pls remove all comments from code before commit and push to repo

File: book_review/app/controllers/book_ratings_controller.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  # GET /book_ratings.json
+  def index
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: I think in project will not be present page with ALL book ratings, you can remove index method, route and views

File: book_review/app/controllers/book_ratings_controller.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  # POST /book_ratings
+  # POST /book_ratings.json
+  def create
+    @book = Book.find(params[:book_id])
+    @book_rating = @book.book_ratings.new(book_rating_params)
+    @book_rating.user = current_user
+    respond_to do |format|
+      if @book_rating.save
+        format.html { redirect_to @book, notice: 'Book rating was successfully created.' }
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: redirect_to @book, notice: 'Book rating was successfully created.'

File: book_review/app/controllers/book_ratings_controller.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      else
+        format.html { render :new }
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: render :new

File: book_review/app/controllers/book_ratings_controller.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      end
+    end
+  end
+
+  # PATCH/PUT /book_ratings/1
+  # PATCH/PUT /book_ratings/1.json
+  def update
+    respond_to do |format|
+      if @book_rating.update(book_rating_params)
+        format.html { redirect_to @book_rating.book, notice: 'Book rating was successfully updated.' }
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: remove "format.html"

File: book_review/app/controllers/book_ratings_controller.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      end
+    end
+  end
+
+  # DELETE /book_ratings/1
+  # DELETE /book_ratings/1.json
+  def destroy
+    @book_rating.destroy
+    respond_to do |format|
+      format.html { redirect_to book_ratings_url, notice: 'Book rating was successfully destroyed.' }
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: remove "format.html"

File: book_review/app/controllers/book_ratings_controller.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_book_rating
+      @book_rating = BookRating.find(params[:id])
+    end
+
+    # Never trust parameters from the scary internet, only allow the white list through.
+    def book_rating_params
+      params.require(:book_rating).permit(:rating_value, :book_id, :user_id)
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: смысл этого permit в том, чтобы определить список атрибутов, которые можно изменять запросами пользователя. Поскольку пользователь вряд ли может выбирать, какому пользователю присвоить book_rating, то user_id тут кажется лишним, думаю, надо убрать

File: book_review/app/controllers/books_controller.rb [hunk 10]
```diff
@@ -11,6 +11,17 @@ def index
...
+    @book = Book.find(params[:id])
+    @book_rating = current_user.book_ratings.find_by(book_id: params[:id])
+    # if @book_rating.nil?
+    #   @book_rating = @book.book_ratings.new
+    # end
+
+    @overall_rating = Book.find(params[:id]).rating
+
+    # @user_book_rating =§
+    # @b_r = @book.book_ratings.new
```
Comment (ghost) [inline book_review/app/controllers/books_controller.rb]: pls remove commented code

File: book_review/app/models/book.rb [hunk 11]
```diff
@@ -1,5 +1,22 @@
 class Book < ApplicationRecord
   belongs_to :user
   has_many :review_comments, dependent: :destroy
+  has_many :book_ratings, dependent: :destroy
+  validates :title, presence: true,
+            length: { minimum: 5}
```
Comment (ghost) [inline book_review/app/models/book.rb]: remove this validation pls

File: book_review/app/models/book_rating.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+class BookRating < ApplicationRecord
+  belongs_to :book
+  belongs_to :user
+
+  validates :user_id, uniqueness: { scope: :book_id }
+
+  validates_numericality_of :rating_value, :greater_than_or_equal_to => 0, :less_than_or_equal_to => 10
```
Comment (ghost) [inline book_review/app/models/book_rating.rb]: validates :rating_value, in: 0..10

File: book_review/app/models/user.rb [hunk 13]
```diff
@@ -1,6 +1,7 @@
 class User < ApplicationRecord
   has_many :books
   has_many :review_comments
+  has_many :book_ratings
```
Comment (ghost) [inline book_review/app/models/user.rb]: dependent: :destroy

File: book_review/app/views/book_ratings/_book_rating.json.jbuilder [hunk 14]
```diff
@@ -0,0 +1,2 @@
+json.extract! book_rating, :id, :rating_value, :book_id, :created_at, :updated_at
```
Comment (ghost) [inline book_review/app/views/book_ratings/_book_rating.json.jbuilder]: not used file
pls remove

File: book_review/app/views/book_ratings/index.html.slim [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  thead
+    tr
+      th Rating value
+      th Book
+      th
+      th
+      th
+
+  tbody
+    - @book_ratings.each do |book_rating|
```
Comment (ghost) [inline book_review/app/views/book_ratings/index.html.slim]: you can place content for each book rating to file _book_rating.html.slim and render it simply with command

= render @book_ratings

File: book_review/app/views/books/show.html.slim [hunk 16]
```diff
@@ -15,7 +17,25 @@ p
...
+
+p
+  strong Overrall rating:
+  = @overall_rating
+p
+  strong Stars:
+  - if @book_rating
+    = @book_rating.rating_value
+  - else
+    = form_for [@book, @book.book_ratings.new] do |f|
```
Comment (ghost) [inline book_review/app/views/books/show.html.slim]: pls put form in separate file

File: book_review/config/routes.rb [hunk 17]
```diff
@@ -3,6 +3,7 @@
   #, only %i{create update destroy}
   resources :books do
     resources :review_comments
+    resources :book_ratings
```
Comment (ghost) [inline book_review/config/routes.rb]: you will not use :index, so pls use :only method to set needed routes

File: book_review/test/controllers/book_ratings_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,48 @@
+require 'test_helper'
```
Comment (ghost) [inline book_review/test/controllers/book_ratings_controller_test.rb]: remove this test files pls

File: book_review/test/fixtures/book_ratings.yml [hunk 19]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (ghost) [inline book_review/test/fixtures/book_ratings.yml]: and this

File: book_review/test/system/book_ratings_test.rb [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  test "destroying a Book rating" do
+    visit book_ratings_url
+    page.accept_confirm do
+      click_on "Destroy", match: :first
+    end
+
+    assert_text "Book rating was successfully destroyed"
+  end
```
Comment (ghost) [inline book_review/test/system/book_ratings_test.rb]: pls remove all from folder "test"

File: book_review/Gemfile [hunk 21]
```diff
@@ -33,6 +33,9 @@ gem 'jbuilder', '~> 2.5'
 # Use Capistrano for deployment
 # gem 'capistrano-rails', group: :development
 
+#for controllers
```
Comment (ghost) [inline book_review/Gemfile]: pls remove this comment, we all know that this is for controllers

File: book_review/app/controllers/book_ratings_controller.rb [hunk 22]
```diff
@@ -0,0 +1,44 @@
+class BookRatingsController < ApplicationController
+  before_action :set_book_rating, only: [:show, :edit, :update, :destroy]
+
+  expose :book_rating
+  expose :book, ->{ Book.find(params[:book_id]) }
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: expose gem will do it itself, you dont need to find by param. leave only:

expose :book

File: book_review/app/controllers/book_ratings_controller.rb [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+
+  def create
+    # byebug
+    book_rating = book.book_ratings.new(book_rating_params)
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: you dont need this line, expose will build object for you

File: book_review/app/controllers/book_ratings_controller.rb [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    if book_rating.save
+      redirect_to book, notice: 'Book rating was successfully created.'
+    else
+      render :new
+    end
+
+  end
+
+  def update
+    book = Book.find(params[:book_id])
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: and this

File: book_review/app/controllers/book_ratings_controller.rb [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+  end
+
+  def destroy
+    book_rating.destroy
+    redirect_to book_ratings_url, notice: 'Book rating was successfully destroyed.'
+  end
+
+  private
+    def set_book_rating
```
Comment (ghost) [inline book_review/app/controllers/book_ratings_controller.rb]: you will not use this method, remove it and before_action

File: book_review/app/controllers/books_controller.rb [hunk 26]
```diff
@@ -1,73 +1,49 @@
...
-  def index
-    @books = Book.all
-  end
-
-  # GET /books/1
-  # GET /books/1.json
-  def show
-  end
+  expose :book
+  expose :books, -> {Book.all}
```
Comment (ghost) [inline book_review/app/controllers/books_controller.rb]: expose :books, -> { current_user.books }
expose :books, parent: :current_user

File: book_review/app/controllers/books_controller.rb [hunk 27]
```diff
@@ -1,73 +1,49 @@
...
+  expose :book_rating, ->{current_user.book_ratings.find_by(book_id: params[:id])}
```
Comment (ghost) [inline book_review/app/controllers/books_controller.rb]: expose :book_rating, -> { book.book_ratings }

File: book_review/app/controllers/books_controller.rb [hunk 28]
```diff
@@ -1,73 +1,49 @@
...
+  expose :overall_rating, ->{Book.find(params[:id]).rating}
```
Comment (ghost) [inline book_review/app/controllers/books_controller.rb]: expose :overall_rating, -> { book.rating }

File: book_review/app/controllers/books_controller.rb [hunk 29]
```diff
@@ -1,73 +1,49 @@
...
 
-  # GET /books/new
-  def new
-    @book = Book.new
 
+  def show
+    unless book_rating
+      book_rating = book.book_ratings.new
```
Comment (ghost) [inline book_review/app/controllers/books_controller.rb]: pls remove it

File: book_review/app/controllers/books_controller.rb [hunk 30]
```diff
@@ -1,73 +1,49 @@
...
+    end
+    overall_rating
```
Comment (ghost) [inline book_review/app/controllers/books_controller.rb]: pls remove this

File: book_review/app/helpers/book_ratings_helper.rb [hunk 31]
```diff
@@ -0,0 +1,2 @@
+module BookRatingsHelper
```
Comment (ghost) [inline book_review/app/helpers/book_ratings_helper.rb]: pls remove

File: book_review/app/models/book.rb [hunk 32]
```diff
@@ -1,5 +1,24 @@
...
+  has_many :book_ratings, dependent: :destroy
+
+  has_attached_file :book_image, styles: {medium: "300x300#", thumb: "100x100>"},
+                    :default_url => "/images/:style/book.png"
+
+  validates_attachment_content_type :book_image, :content_type => /\Aimage\/.*\Z/
+
+
+  def rating
+    if book_ratings.count == 0
```
Comment (ghost) [inline book_review/app/models/book.rb]: return unless book_ratings.any?

book_ratings.sum(:rating_value) / book_ratings.count

File: book_review/app/views/book_ratings/index.json.jbuilder [hunk 33]
```diff
@@ -0,0 +1 @@
+json.array! @book_ratings, partial: 'book_ratings/book_rating', as: :book_rating
```
Comment (ghost) [inline book_review/app/views/book_ratings/index.json.jbuilder]: remove

File: book_review/app/views/book_ratings/show.json.jbuilder [hunk 34]
```diff
@@ -0,0 +1 @@
+json.partial! "book_ratings/book_rating", book_rating: @book_rating
```
Comment (ghost) [inline book_review/app/views/book_ratings/show.json.jbuilder]: pls remove

File: book_review/app/controllers/book_ratings_controller.rb [hunk 35]
```diff
@@ -0,0 +1,39 @@
+class BookRatingsController < ApplicationController
+  expose :book_rating
+  expose :book
+
+  def create
+    book_rating = book.book_ratings.new(book_rating_params)
```
Comment (camelot729) [inline book_review/app/controllers/book_ratings_controller.rb]: book_rating.book = book

File: book_review/app/controllers/books_controller.rb [hunk 36]
```diff
@@ -1,73 +1,54 @@
...
 
-  # POST /books
-  # POST /books.json
   def create
-    @book = current_user.books.new(book_params)
-
-
-    if @book.save
-      redirect_to @book, notice: 'Book was successfully created.'
+    book = current_user.books.new(book_params)
```
Comment (camelot729) [inline book_review/app/controllers/books_controller.rb]: book.user = current_user

File: book_review/app/controllers/books_controller.rb [hunk 37]
```diff
@@ -1,73 +1,54 @@
...
-  # PATCH/PUT /books/1.json
   def update
-    @user_book = current_user.book.find(params[:id])
-    respond_to do |format|
-      if @book.update(book_params)
-        redirect_to @book, notice: 'Book was successfully updated.'
-      else
-        render :edit
-      end
+    book = current_user.book.find(params[:id])
```
Comment (camelot729) [inline book_review/app/controllers/books_controller.rb]: the  same

File: book_review/app/controllers/review_comments_controller.rb [hunk 38]
```diff
@@ -1,75 +1,43 @@
...
-  end
-
-  # GET /review_comments/1/edit
-  def edit
-  end
-
-  # POST /review_comments
-  # POST /review_comments.json
   def create
+    review_comment = book.review_comments.create(review_comment_params)
```
Comment (camelot729) [inline book_review/app/controllers/review_comments_controller.rb:7]: review_comment.book = book

File: book_review/app/controllers/users_controller.rb [hunk 39]
```diff
@@ -1,44 +1,33 @@
 class UsersController < ApplicationController
-
   before_action :authenticate_user!
-
-  def index
-  end
-
-  def edit
-  end
+  expose :user, -> { current_user }
```
Comment (camelot729) [inline book_review/app/controllers/users_controller.rb:3]: remove

**Pull request merged**